gcc/emit-rtl.c

```diff
@@ -2456,8 +2457,13 @@ adjust_address_1 (rtx memref, machine_mode mode, poly_int64 offset,
 
   if (maybe_ne (size, 0))
     {
+      /* Use size of record or union type to check object bounds correctly */
```

Comment:
However this does not cover case of flexible array members. Probably we can't make object boundary check in case of structures or records.
```
if (adjust_object && maybe_gt (offset + size, attrs.size) && !(attrs.expr && RECORD_OR_UNION_TYPE_P (TREE_TYPE (attrs.expr))))
```